File: python/underScorifySubString.py

# Time -> 0(N+M) Space -> 0(N)
def underscorifySubstring(string, substring):
    locations = []
    i = 0
    while i < len(string):
        val = string.find(substring, i)
        if val != -1:
            locations.append([val, val+len(substring)])
            i = val+1
        else:
            break

    if len(locations) == 0:
        return string
    
    newlocations = [locations[0]]
    for current in locations[1:]:
        last = newlocations[-1]
        
        if current[0] <= last[1]:
            last[1] = current[1]
        else:
            newlocations.append(current)

    print(newlocations, locations)
    
    newString = []
    i, j = 0, 0
    while i < len(string):
        if j < len(newlocations) and i == newlocations[j][0]:
            newString.append("_"+string[newlocations[j][0]:newlocations[j][1]]+"_")
            i = newlocations[j][1]
            j += 1
            continue
        else:
            newString.append(string[i])
        i += 1
            
    return "".join(newString)
    
print(underscorifySubstring("testtestesttestest this is a test to see if it works", "test"))

File: go/underScorifySubString/underScorifySubString.go
package main

import (
	"fmt"
	"strings"
)

// Time -> 0(N+M) Space -> 0(N)
func UnderScorifySubString(str, substring string) string {
	var locations [][]int
	var i int
	for i < len(str) {
		idx := strings.Index(str[i:], substring)
		if idx == -1 {break}
		start := i + idx
		end := start + len(substring)
		locations = append(locations, []int{start, end})
		i = start + 1
	}

	if len(locations) == 0 {return str}
	newlocation := [][]int{locations[0]}
	for _, val := range locations[1:] {
		last := newlocation[len(newlocation)-1]

		if val[0] <= last[1] {
			last[1] = val[1]
		}else {newlocation = append(newlocation, val)}
	}

	i = 0
	var j int
	var result strings.Builder
	for i < len(str) {
		if j < len(newlocation) && i == newlocation[j][0] {
			result.WriteString("_"+str[i:newlocation[j][1]]+"_")
			i = newlocation[j][1]
			j += 1
		}else {
			result.WriteString(string(str[i]))
			i++
		}
	}
	return result.String()
}

func main() {
	fmt.Println(UnderScorifySubString("testesttesttestest is not going well", "test"))
}